str.lisp

```diff
@@ -137,27 +151,126 @@
 (defun version ()
   (print +version+))
 
+;; as a dedicated condition to help detect it in tests
+;; (up to 0.18.1 a negative index would mean zero)
+(define-condition negative-index (warning) ()
+  (:report
+   "STR possible breaking change: negative index means 'from end' now"))
+
+(declaim (inline index))
+
+(defun index (string/length index)
+  ""
+  (declare (optimize (speed 3))
+           #+sbcl ;; inlining: notes in caller code
+           (sb-ext:muffle-conditions sb-ext:compiler-note)
+           (type (or null fixnum) index)
+           (type (or vector array-total-size) string/length))
+  (let ((length (etypecase string/length
+                  (string (length string/length))
+                  (array-total-size string/length))))
+    (cond
+      ((member index '(nil t)) length)
+      ((= 0 length) 0)
+      (t
+       (locally (declare (type (and fixnum (not (eql 0))) length))
+         (when (< index 0)
+           (warn 'negative-index))
+         (multiple-value-bind (quotient mod) (floor index length)
+           (if (<= -1 quotient 0)
+               ;; wrap when index is between -length and length
+               mod
+               ;; otherwise clamp to either side
+               (if (< quotient 0) 0 length))))))))
+
+(defmacro with-indices ((&rest indices) string &body body)
+  "Ensure INDICES are computed according to INDEX w.r.t. STRING in BODY.
+
+INDICES is a list of either a symbol NAME or a couple (NAME VALUE); if
+only NAME is provided, NAME is expected to be bound in the current
+context to the VALUE associaed with NAME.
+
+Inside BODY, each NAME is bound by LET* to (INDEX STRING VALUE).
+"
+  (let (($string (gensym)))
+    `(let ((,$string ,string))
+       ,(flet ((make-binding (index)
+                 (let ((index (if (listp index) index (list index))))
+                   (destructuring-bind (name &optional value) index
+                     (let ((index-expr (or value name)))
+                       `(,name (index ,$string ,index-expr)))))))
+          `(let* ,(mapcar #'make-binding indices)
+             ,@body)))))
+
+(declaim (inline slice%))
+(defun slice% (start end s sharedp)
+  #+sbcl ;; inlining: notes in caller code
+  (declare (sb-ext:muffle-conditions sb-ext:compiler-note))
+  (let ((length (- end start)))
+    (cond
+      ((<= length 0) "")
+      ((not sharedp) (subseq s start end))
+      (t (make-array length
+                     :element-type (array-element-type s)
+                     :displaced-to s
+                     :displaced-index-offset start)))))
+
+(defun slice (start end s &optional (sharedp *sharedp*))
+  (when s
+    (with-indices ((start (or start 0)) end) s
```

Comment:
This is less about saving lines and more about having a descriptive name for what is being computed. Whenever there is code that relies on a string index we can wrap the existing code with this.
Currently this is equivalent to:

    (let* ((start (index s (or start 0)))
             (end (index s end)))
        ....)

But this may evolve later to do things better, without having to update the call (e.g. add declarations)